cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -180,25 +183,72 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
 	return strings.TrimPrefix(cert.Subject.CommonName, constants.NodesUserPrefix), nil
 }
 
+// getAPIEndpointOrMigrate returns the APIEndpoint for the current node as reported by getAPIEndpoint. If this
+// information is not yet present, chances are that we haven't yet migrated this information to reside on
+// pod annotations, fallback in that case to GetClusterStatus, that will compose this information out of the
+// ClusterStatus struct present in the kubeadm-config ConfigMap.
+//
+// TODO: remove after 1.20, when the ClusterStatus struct is removed from the kubeadm-config ConfigMap;
+//       callers of this function should call getAPIEndpoint directly when that happens.
+func getAPIEndpointOrMigrate(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
+	apiEndpoint_ := kubeadmapi.APIEndpoint{}
+	if err := getAPIEndpoint(client, nodeName, &apiEndpoint_); err == nil {
+		*apiEndpoint = apiEndpoint_
+		return nil
+	}
+	clusterStatus, err := GetClusterStatus(client)
+	if err != nil {
+		return errors.Wrap(err, "could not retrieve cluster status")
+	}
+	if apiEndpoint_, ok := clusterStatus.APIEndpoints[nodeName]; ok {
+		*apiEndpoint = apiEndpoint_
+		return nil
+	}
+	return errors.Errorf("could not find node name %q in cluster status", nodeName)
+}
+
 // getAPIEndpoint returns the APIEndpoint for the current node
-func getAPIEndpoint(data map[string]string, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
-	// gets the ClusterStatus from kubeadm-config
-	clusterStatus, err := UnmarshalClusterStatus(data)
+func getAPIEndpoint(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
+	apiEndpoints, err := GetAllAPIEndpoints(client)
```

Comment:
Ideally, we don't need to fetch all endpoints. You can just add a field selector for `spec.nodeName=<nodeName>` to the query. That will get you a list, which should contain only a single pod object.
We don't seem to be using `GetAllAPIEndpoints` anywhere else so we can certainly go without it.

Reply:
Thought that this could come handy at some point, after all since we already did one request, getting more information with a single request is "free", except for some local cycles composing the information.

Reply:
Will update nevertheless.